iot_packet_verifier/src/verifier.rs

```diff
@@ -0,0 +1,496 @@
+    type Error;
+
+    async fn fetch_org(
+        &mut self,
+        oui: u64,
+        cache: &mut HashMap<u64, PublicKeyBinary>,
+    ) -> Result<PublicKeyBinary, Self::Error>;
+
+    async fn enable_org(&mut self, oui: u64) -> Result<(), Self::Error>;
+
+    async fn disable_org(&mut self, oui: u64) -> Result<(), Self::Error>;
+}
+
+// TODO: Move this somewhere else
+
+// Probably should change name to something like OrgClientCache to be more
+// consistent with BalanceCache
+pub struct CachedOrgClient {
+    pub keypair: Keypair,
+    pub enabled_clients: HashMap<u64, bool>,
+    pub client: OrgClient<Channel>,
+}
+
+impl CachedOrgClient {
+    pub fn new(client: OrgClient<Channel>, keypair: Keypair) -> Self {
+        CachedOrgClient {
+            keypair,
+            enabled_clients: HashMap::new(),
+            client,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum OrgClientError {
+    #[error("Rpc error: {0}")]
+    RpcError(#[from] tonic::Status),
+    #[error("Crypto error: {0}")]
+    CryptoError(#[from] helium_crypto::Error),
+}
+
+#[async_trait]
+impl ConfigServer for CachedOrgClient {
+    type Error = OrgClientError;
+
+    async fn fetch_org(
+        &mut self,
+        oui: u64,
+        cache: &mut HashMap<u64, PublicKeyBinary>,
+    ) -> Result<PublicKeyBinary, Self::Error> {
+        if let Entry::Vacant(e) = cache.entry(oui) {
+            let req = OrgGetReqV1 { oui };
+            let pubkey =
+                PublicKeyBinary::from(self.client.get(req).await?.into_inner().org.unwrap().owner);
+            e.insert(pubkey);
+        }
+        Ok(cache.get(&oui).unwrap().clone())
+    }
+
+    async fn enable_org(&mut self, oui: u64) -> Result<(), Self::Error> {
+        if !mem::replace(self.enabled_clients.entry(oui).or_insert(false), true) {
+            let mut req = OrgEnableReqV1 {
+                oui,
+                timestamp: Utc::now().timestamp_millis() as u64,
+                signature: vec![],
+            };
+            let signature = self.keypair.sign(&req.encode_to_vec())?;
+            req.signature = signature;
+            let _ = self.client.enable(req).await?;
+        }
+        Ok(())
+    }
+
+    async fn disable_org(&mut self, oui: u64) -> Result<(), Self::Error> {
+        if mem::replace(self.enabled_clients.entry(oui).or_insert(true), false) {
+            let mut req = OrgDisableReqV1 {
+                oui,
+                timestamp: Utc::now().timestamp_millis() as u64,
+                signature: vec![],
+            };
+            let signature = self.keypair.sign(&req.encode_to_vec())?;
+            req.signature = signature;
+            let _ = self.client.disable(req).await?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+pub trait Burner {
+    type Error;
+
+    async fn burn(&mut self, payer: &PublicKeyBinary, amount: u64) -> Result<(), Self::Error>;
+}
+
+#[async_trait]
+impl Burner for &'_ mut Transaction<'_, Postgres> {
```

Comment:
i'm still not a fan of the overloading of the name `Burner` since we've got a `Burner` struct that routinely does burn operations and then we've got a `Burner` trait in the same crate, both of which define a `burn()` function that writes to the same database table but other than the table there's nothing that actually links these two parts of the dc burn process so i think this could be ripe for a refactor bug in the future with these being defined in separate files/modules and the names implying they're more related than two separate processes writing to the same table.

at the very least i think this should be renamed to something distinct from the `Burner` struct in the `burner.rs` module like `BurnMeter` with a `add_burn()` function so long as this whole thing can't be threaded through the `Burner` struct/process itself and have _all_ of the account burn access to the database be managed by a single process (such as with a message channel or similar)

Reply:
You're right, I'll rename this to `PendingBurns`